### ISSUE TITLE
Make IF great: Improve interface and include iterative method 

### ISSUE THREAD

**Issue by Xuzzo** [appliedAI-Initiative/pyDVL#311]
There are several changes that are necessary for IF to be great. In order, the following changes must be implemented:
- [x] Conjugate gradient must be implemented within the framework. This is necessary for managing device computation. In the current setup, torch handles all the differentiation, which typically happens in a gpu, but then for hessian inversion data is offloaded to the cpu, causing massive overhead. 
- [x] Influence factors computation should be done in a separate method, so that the result can be stored and re-used. At the moment, influence factors are calculated internally in compute_influences and are not cached.
- [x] The stochastic Hessian inversion (also known as LISSA, Linear time Stochastic Second-Order Algorithm) also needs to be implemented, both for paper reproductions and because it is very useful for large models. 
- [x] Currently code expects all training and test data to be fully loaded into RAM. Include options for pytorch DataLoaders
- [x] With DataLoaders, it will be important to handle the device properly. Check that everything continues to work on a gpu

For now, all these things should be done in Pytorch, keeping in mind that the library aims to be eventually framework independent.

**Comment by mdbenito** [comment]
@Xuzzo This is addressed in your PRs, right? Maybe check off the items?

**Comment by mdbenito** [comment]
This was done in #339 and #306
